txp/Pos/Txp/Logic/Local.hs

```diff
@@ -77,68 +75,49 @@ txProcessTransaction itw@(txId, txAux) = do
-            bad :: [(StakeholderId, Coin)] -> Bool
-            bad (map fst -> txOutDistr) =
-                -- Has unrelated address
-                any (not . inBoot) txOutDistr ||
-                -- Not all genesis boot addrs are mentioned
-                any (not . mentioned txOutDistr) (HS.toList bootHolders)
-        in NE.filter bad txDistr
-
     processTxDo
-        :: Utxo
-        -> HashSet StakeholderId
-        -> ToilEnv
+        :: EpochIndex
+        -> BlockVersionData
+        -> GenesisStakeholders
+        -> Utxo
         -> HeaderHash
         -> (TxId, TxAux)
-        -> Bool
         -> TxpLocalDataPure
         -> (Either ToilVerFailure (), TxpLocalDataPure)
-    processTxDo resolved bootHolders toilEnv tipDB tx bootEra txld@(uv, mp, undo, tip, ()) = do
-        let tipMismatch = tipDB /= tip
-            bootRel = notBootRelated bootHolders
-        if | tipMismatch ->
-             (Left $ ToilTipsMismatch tipDB tip, txld)
-           | bootEra && not (null bootRel) ->
-            -- We do not allow txs with non-boot-addr stake distr in boot era.
-             (Left $ ToilBootDifferentStake $ unsafeHead bootRel, txld)
-           | otherwise ->
-             let res = (runExceptT $
-                        flip evalUtxoStateT resolved $
-                        execToilTLocal uv mp undo $
-                        processTx tx) toilEnv
-             in case res of
-                 Left er  -> (Left er, txld)
-                 Right ToilModifier{..} ->
-                     (Right (), (_tmUtxo, _tmMemPool, _tmUndos, tip, ()))
+    processTxDo curEpoch bvd genStks resolved tipDB tx txld@(uv, mp, undo, tip, ())
+        | tipDB /= tip =
+            (Left $ ToilTipsMismatch tipDB tip, txld)
+        | otherwise =
+            let res = (runExceptT $
+                    flip runReaderT genStks $
+                    flip evalUtxoStateT resolved $
+                    execToilTLocal uv mp undo $
+                    processTx curEpoch tx) bvd
+            in case res of
+                Left er  -> (Left er, txld)
+                Right ToilModifier{..} ->
+                    (Right (), (_tmUtxo, _tmMemPool, _tmUndos, tip, ()))
 
 -- | 1. Recompute UtxoView by current MemPool
 -- | 2. Remove invalid transactions from MemPool
 -- | 3. Set new tip to txp local data
-txNormalize ::
-       ( MonadIO m
-       , MonadBaseControl IO m
-       , MonadDBRead m
-       , MonadGState m
-       , MonadTxpMem () ctx m
-       , WithLogger m
-       )
+txNormalize
+    :: ( TxpLocalWorkMode ctx m
+       , MonadSlots m)
     => m ()
-txNormalize = do
-    utxoTip <- GS.getTip
-    localTxs <- getLocalTxs
-    ToilModifier {..} <-
-        runDBToil $ execToilTLocal mempty def mempty $ normalizeToil localTxs
-    setTxpLocalData "txNormalize" (_tmUtxo, _tmMemPool, _tmUndos, utxoTip, _tmExtra)
+txNormalize = getCurrentSlot >>= \case
+    Nothing -> clearTxpMemPool "txNormalize "
```

Comment:
It means that we delete all transactions from the mempool if we don't know the current slot. Which is fine because it means we are lost. What I don't like is that we don't update tip. It's bad @pva701